src/Context/UserInfo.tsx

```diff
@@ -0,0 +1,162 @@
+import { createContext, useContext, useReducer } from "react";
+
+interface ICar {
+  id: string,
+  brand: string,
+  color: string
+}
+
+interface IState {
+  name: string;
+  lastName: string;
+  isLoggedIn: boolean,
+  cars: ICar[],
+}
+
+interface IContext {
+  name: string;
+  lastName: string;
+  isLoggedIn: boolean,
+  cars: ICar[],
+  toggleLogin: () => void,
+  addCar: (e: React.SyntheticEvent) => void;
+  deleteCar: (id: string) => void;
+  editCar: (e: React.SyntheticEvent) => void;
+}
+
+interface IProps {
+  children: React.ReactChild;
+}
+
+type IReducerAction = {
+  type: ReducerActionType.ADD_CAR;
+  data: ICar;
+} | {
+  type: ReducerActionType.DELETE_CAR;
+  data: string;
+} | {
+  type: ReducerActionType.EDIT_CAR;
+  data: ICar;
+} | {
+  type: ReducerActionType.LOGGED;
+};
+
+enum ReducerActionType {
+  ADD_CAR,
+  DELETE_CAR,
+  EDIT_CAR,
+  LOGGED,
+}
+
+export const UserContext = createContext<IContext | null>(null);
+
+const initialState: IState = {
+  name: "Gerardo",
+  lastName: "Estrada",
+  isLoggedIn: false,
+  cars: [
+    {
+      id: "1",
+      brand: "Toyota",
+      color: "Blue"
+    },
+    {
+      id: "2",
+      brand: "Nissan",
+      color: "Red"
+    }
+  ]
+};
+
+export const useUserContext = () => {
+  const context = useContext(UserContext);
+  if (!context) {
+    throw "You can't use the userContext outside of the provider";
+  }
+  return context;
+};
+
+const reducerCars = (state: IState, action: IReducerAction): IState => {
+  switch (action.type) {
+    case ReducerActionType.ADD_CAR:
+      return { ...state, cars: [...state.cars, action.data] };
+
+    case ReducerActionType.EDIT_CAR:
+      const updateCar = () => {
+        const newData: ICar[] = new Array(state.cars.length)
+
+        for (let i = 0; i < newData.length; i++) {
+          state.cars[i].id === action.data.id ? newData[i] = action.data : newData[i] = state.cars[i]
+        }
+
+        return newData
+      }
+
+      return { ...state, cars: updateCar() }
+
+    case ReducerActionType.DELETE_CAR:
+      return { ...state, cars: [...state.cars.filter((car) => car.id !== action.data)] };
```

Comment:
I don't think you need to spread the filtered array, you can just assign it to `car`.
```suggestion
      return { ...state, cars: state.cars.filter((car) => car.id !== action.data)};
```